compiler/rustc_span/src/hygiene.rs

```diff
@@ -1272,3 +1305,116 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
         panic!("cannot decode `SyntaxContext` with `{}`", std::any::type_name::<D>());
     }
 }
+
+/// Updates the `disambiguator` field of the corresponding `ExpnData`
+/// such that the `Fingerprint` of the `ExpnData` does not collide with
+/// any other `ExpnIds`.
+///
+/// This method is called only when an `ExpnData` is first associated
+/// with an `ExpnId` (when the `ExpnId` is initially constructed, or via
+/// `set_expn_data`). It is *not* called for foreign `ExpnId`s deserialized
+/// from another crate's metadata - since `ExpnData` includes a `krate` field,
+/// collisions are only possible between `ExpnId`s within the same crate.
+pub(super) fn update_disambiguator(expn_id: ExpnId) {
```

Comment:
```suggestion
fn update_disambiguator(expn_id: ExpnId) {
```

Reply:
This also should be a method on `HygieneData` if possible.
Inside this functions we take the hygiene data lock (e.g. in `expn_id.expn_data()`), but that is extra work - at call points of `update_disambiguator` we already have to take that lock.

Reply:
We cannot call this method with the hygiene data lock taken, since hashing the `ExpnData` needs to take the lock.